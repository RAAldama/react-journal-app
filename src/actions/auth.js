import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { uiFinishLoading, uiStartLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return ( dispatch ) => {
        dispatch( uiStartLoading() );

        firebase.auth().signInWithEmailAndPassword(email, password).then(
            ({ user }) => {
                dispatch( login(user.uid, user.displayName) );
                dispatch( uiFinishLoading() )
            }
        ).catch(
            e => {
                console.log(e);
                dispatch( uiFinishLoading() );
            }
        )
    }
}

export const startRegisterEmailPassword = (email, password, username) => {
    return ( dispatch ) => {
        firebase.auth().createUserWithEmailAndPassword(email, password).then(
            async( {user} ) => {
                await user.updateProfile({ displayName: username });

                dispatch( login(user.uid, user.displayName) );
            }
        ).catch(
            e => console.log(e)
        )
    }
}

export const startLoginGoogle = () => {
    return ( dispatch ) => {
        firebase.auth().signInWithPopup( googleAuthProvider ).then(
            ( {user} ) => {
                dispatch( login(user.uid, user.displayName) );
            }
        );
    }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}