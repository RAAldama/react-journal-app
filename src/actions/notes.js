import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    return async(dispatch, getSate) => {

        const uid = getSate().auth.uid;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await db.collection(`${uid}/journal/notes`).add( newNote );
        dispatch( activeNote(doc.id, newNote) );

    }
}

export const activeNote = (id, note) => {
    return {
        type: types.notesActive,
        payload: {id, ...note}
    }
}

export const startLoginNotes = (uid) => {
    return async (dispatch) => {

        const notes = await loadNotes(uid);
        dispatch ( setNotes(notes) );
        
    }
}

export const setNotes = (notes) => {
    return {
        type: types.notesLoad,
        payload: notes
    }
}

export const startSaveNote = (note) => {
    return async(dispatch, getSate) => {

        const uid = getSate().auth.uid;

        if(!note.url){
            delete note.url;
        }

        const note2Firestore = {...note};

        delete note2Firestore.id;

        await db.doc(`${uid}/journal/notes/${note.id}`).update(note2Firestore);

        dispatch( refreshNote(note.id, note) );
        Swal.fire('Nota salvada', note.title, 'success');

    }
}

export const refreshNote = (id, note) => {
    return {
        type: types.notesUpdated,
        payload: { id, note }
    }
}