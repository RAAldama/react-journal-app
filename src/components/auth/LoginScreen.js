import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";

import { startLoginEmailPassword, startLoginGoogle } from "../../actions/auth";
import { useForm } from "../../hooks/useForm";

export const LoginScreen = () => {

  const dispatch = useDispatch()
  const { loading } = useSelector( state => state.ui );

  const [ formValues, handleInputChange ] = useForm({
    email: '',
    password: ''
  });

  const {email, password} = formValues;

  const handleLogin = (e) => {
    e.preventDefault();

    dispatch( startLoginEmailPassword(email, password) );
  }

  const handleGoogleLogin = () => {
    dispatch( startLoginGoogle() );
  }

  return (
    <>
      <h3 className="auth__title">Login</h3>

      <form className="animate__animated animate__fadeIn" onSubmit={handleLogin}>
        <input 
          className="auth__input" 
          type="text" 
          placeholder="Correo electrónico" 
          name="email" 
          value={email} 
          onChange={handleInputChange}
          autoComplete="off" 
        />
        <input 
          className="auth__input" 
          type="password" 
          placeholder="Contraseña" 
          name="password" 
          value={password} 
          onChange={handleInputChange}
          />

        <button type="submit" className="btn btn-primary btn-block" disabled={ loading }>
          Ingresar
        </button>

        <div className="auth__social-networks">
          <p>Unete con tus redes sociales :D</p>

          <div className="google-btn" onClick={ handleGoogleLogin }>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>

            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link className="link" to="/auth/register">
          ¿No tienes cuenta? ¡Crea una aquí!
        </Link>
        
      </form>
    </>
  );
};
