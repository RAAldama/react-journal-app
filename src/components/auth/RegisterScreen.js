import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import validator from "validator";

import { startRegisterEmailPassword } from "../../actions/auth";
import { removeError, setError } from "../../actions/ui";
import { useForm } from "../../hooks/useForm";

export const RegisterScreen = () => {

  const dispatch = useDispatch();
  const { msgError } = useSelector( state => state.ui );

  const [formValues, handleInputChange] = useForm({
    name: 'albus',
    email: 'lel@lel.com',
    password: '123456',
    password2: '123456',
  });

  const {name, email, password, password2} = formValues;

  const handleRegister = (e) => {
    e.preventDefault();

    if( isFormValid() ){
      dispatch( startRegisterEmailPassword(email, password, name) );
    }
  }

  const isFormValid = () => {
    if( name.trim().length === 0 ){
      const error = 'El NOMBRE es obligatorio';
      dispatch( setError(error) );

      return false;

    } else if( !validator.isEmail(email) ) {
      const error = 'El CORREO debe de ser válido';
      dispatch( setError(error) );

      return false;

    } else if( password !== password2 || password.length < 5 ){
      const error = 'Revise la CONTRASEÑA';
      dispatch( setError(error) );

      return false;
    }

    dispatch( removeError() );

    return true;
  }

  return (
    <>
      <h3 className="auth__title">Register</h3>

      <form className="animate__animated animate__fadeIn" onSubmit={handleRegister}>
        
        {
          msgError &&
          <div className="auth__alert-error">
            { msgError }
          </div>
        }

        <input
          className="auth__input"
          type="text"
          placeholder="Usuario"
          name="name"
          value={name} 
          onChange={handleInputChange}
          autoComplete="off"
        />
        <input
          className="auth__input"
          type="text"
          placeholder="Correo electrónico"
          name="email"
          value={email} 
          onChange={handleInputChange}
          autoComplete="off"
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Contraseña"
          name="password"
          value={password} 
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Confirmar contraseña"
          name="password2"
          value={password2} 
          onChange={handleInputChange}
        />

        <button type="submit" className="btn btn-primary btn-block mb-5">
          Registrar
        </button>

        <Link className="link" to="/auth/login">
          ¿Con cuenta? ¡Ingresa aquí!
        </Link>
      </form>
    </>
  );
};
