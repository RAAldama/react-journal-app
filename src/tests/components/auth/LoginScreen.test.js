import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { MemoryRouter } from 'react-router-dom'
import { mount } from "enzyme"
import { Provider } from "react-redux";

import { LoginScreen } from "../../../components/auth/LoginScreen"
import { startLoginEmailPassword, startLoginGoogle } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
    startLoginGoogle: jest.fn(),
    startLoginEmailPassword: jest.fn('', '')
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
}

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount( 
    <Provider store={ store }>
        <MemoryRouter>

            <LoginScreen /> 

        </MemoryRouter>
    </Provider>
);

describe('tests on <LoginScreen />', () => {

    beforeEach(() => {
        store = mockStore(initState);
        jest.clearAllMocks();
    })
    
    test('should show correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('should begin the action of startLoginGoogle', () => {
        wrapper.find('.google-btn').prop('onClick')();

        expect( startLoginGoogle ).toHaveBeenCalled();
    })
    
    test('should begin the action of startLoginEmailPassword', () => {
        wrapper.find('form').prop('onSubmit')({ preventDefault(){} });

        expect( startLoginEmailPassword ).toHaveBeenCalledWith('','');
    })

})
