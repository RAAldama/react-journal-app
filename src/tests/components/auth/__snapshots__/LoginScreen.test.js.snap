// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <LoginScreen /> should show correctly 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Login
  </h3>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Correo electrónico"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Contraseña"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Ingresar
    </button>
    <div
      className="auth__social-networks"
    >
      <p>
        Unete con tus redes sociales :D
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
    </div>
    <a
      className="link"
      href="/auth/register"
      onClick={[Function]}
    >
      ¿No tienes cuenta? ¡Crea una aquí!
    </a>
  </form>,
]
`;
